### PR TITLE
feature: [PHP8.2] Support for readonly classes

### LINKED ISSUE #6604: [PHP8.2] Support for `readonly` classes
The goal of this task is to provide support for [`readonly` classes](https://php.watch/versions/8.2/readonly-classes). New class-level keyword must not break any fixers and should be supported only for PHP 8.2+

### PR DESCRIPTION
**Pull request by Wirone**
Fixes #6604.

These are changes initially proposed by SpacePossum in #6624 but abandoned due to personal reasons. Since it's accepted by the author to use these changes, I've just applied the full patch, so we can continue work.

### CODE REVIEW

**Comment by Wirone** [comment]
@keradus this is only patch applied from @SpacePossum's commit [taken from here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/commit/fb4a4298334be5cbf54bf80c1fbd224c9eee903f.patch), no new changes. I wanted to check if it passes on newest `master` and provide changes if required. At this point it is 1:1 with #6624 🙂

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 1]
```diff
@@ -2385,6 +2385,80 @@ public static function provideIsSuperGlobalCases(): array
...
+            '<?php /* comment */ abstract class Foo {}',
+        ];
+    }
+
+    /**
+     * @requires PHP 8.2
+     *
+     * @dataProvider provideGetClassyModifiersOnPhp82Cases
+     *
+     * @param array<string, int> $expectedModifiers
```
Comment (kubawerlos) [inline tests/Tokenizer/TokensAnalyzerTest.php:2421]:
```suggestion
     * @param array<string, null|int> $expectedModifiers
```

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 2]
```diff
@@ -2385,6 +2385,80 @@ public static function provideIsSuperGlobalCases(): array
         return $cases;
     }
 
+    /**
+     * @dataProvider provideGetClassyModifiersCases
+     *
+     * @param array<string, int> $expectedModifiers
```
Comment (kubawerlos) [inline tests/Tokenizer/TokensAnalyzerTest.php:2391]:
```suggestion
     * @param array<string, null|int> $expectedModifiers
```

File: tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php [hunk 3]
```diff
@@ -327,4 +332,26 @@ public static function provideFix80Cases(): iterable
...
+    {
+        $this->fixer->configure($config);
+        $this->doTest($expected, $input);
+    }
+
+    public function provideFix82Cases(): iterable
+    {
+        yield [
+            '<?php final readonly class A{}',
+            null,
```
Comment (kubawerlos) [inline tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:353]:
```suggestion
            '<?php readonly class A{}',
```

**Review by kubawerlos — CHANGES_REQUESTED** [review]

**Comment by Wirone** [comment]
FYI: Sorry for lack of response, I did not have time lately, but I'll try to look at it later today 🙂

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 4]
```diff
@@ -2385,6 +2385,80 @@ public static function provideIsSuperGlobalCases(): array
...
+     */
+    public function testGetClassyModifiersOnPhp82(array $expectedModifiers, int $index, string $source): void
+    {
+        $tokens = Tokens::fromCode($source);
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
+
+        static::assertSame($expectedModifiers, $tokensAnalyzer->getClassyModifiers($index));
+    }
+
+    public function provideGetClassyModifiersOnPhp82Cases(): iterable
```
[the remark this answers is not in the document]
Comment (Wirone) [reply tests/Tokenizer/TokensAnalyzerTest.php:2431]: Sure, added some more test cases 👍

**Review by kubawerlos — APPROVED** [review]

**Comment by keradus** [comment]
big thanks @SpacePossum and @Wirone , as well as @kubawerlos for nice review!

**Comment by SpacePossum** [comment]
nice one @Wirone :+1:  :tada:

**Comment by Wirone** [comment]
To be fair, it's all thanks to @SpacePossum - I only did maintenance work to refresh the branch and fix little issues to align the code with current CS rules. Anyway, happy that `readonly` classes support got merged 🎉

**Pull request merged**